### PR TITLE
Refactor staking ledger

### PR DESCRIPTION
**Pull request by gpestana**
Moves https://github.com/paritytech/substrate/pull/14582 into `polkadot-sdk`

---

This PR refactors the staking ledger logic to encapsulate all reads and mutations of `Ledger`, `Bonded`, `Payee` and stake locks within the `StakingLedger` struct implementation.

With these changes, all the reads and mutations to the `Ledger`, `Payee` and `Bonded` storage map should be done through the methods exposed by StakingLedger to ensure the data and lock consistency of the operations. The new introduced methods that mutate and read Ledger are:

- `ledger.update()`: inserts/updates a staking ledger in storage; updates staking locks accordingly (and ledger.bond(), which is synthatic sugar for ledger.update())
- `ledger.kill()`: removes all Bonded and StakingLedger related data for a given ledger; updates staking locks accordingly;
`StakingLedger::get(account)`: queries both the `Bonded` and `Ledger` storages and returns a `Option<StakingLedger>`. The pallet impl exposes fn ledger(account) as synthatic sugar for `StakingLedger::get(account)`.

Retrieving a ledger with `StakingLedger::get()` can be done by providing either a stash or controller account. The input must be wrapped in a `StakingAccount` variant (Stash or Controller) which is treated accordingly. This simplifies the caller API but will eventually be deprecated once we completely get rid of the controller account in staking. However, this refactor will help with the work necessary when completely removing the controller.

Other goals:

- No logical changes have been introduced in this PR;
- No breaking changes or updates in wallets required;
- No new storage items or need to perform storage migrations;
- Centralise the changes to bonds and ledger updates to simplify the OnStakingUpdate updates to the target list (related to https://github.com/paritytech/polkadot-sdk/issues/443)

Note: it would be great to prevent or at least raise a warning if `Ledger<T>`, `Payee<T>` and `Bonded<T>` storage types are accessed outside the `StakingLedger` implementation. This PR should not get blocked by that feature, but there's a tracking issue here https://github.com/paritytech/polkadot-sdk/issues/149

Related and step towards https://github.com/paritytech/polkadot-sdk/issues/443

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-additional-tests
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3635717

File: substrate/frame/staking/src/pallet/impls.rs [hunk 1]
```diff
@@ -163,10 +178,15 @@ impl<T: Config> Pallet<T> {
...
-		let controller = Self::bonded(&validator_stash).ok_or_else(|| {
-			Error::<T>::NotStash.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))
+		let account = StakingAccount::Stash(validator_stash.clone());
+		let mut ledger = Self::ledger(account.clone()).or_else(|_| {
+			if StakingLedger::<T>::is_bonded(account) {
+				Err(Error::<T>::NotController.into())
+			} else {
+				Err(Error::<T>::NotStash.with_weight(T::WeightInfo::payout_stakers_alive_staked(0)))
+			}
 		})?;
```
Comment (Ank4n) [inline substrate/frame/staking/src/pallet/impls.rs:188]: wdyt? I think we can return same weight in both errors.
```suggestion
		let mut ledger = Self::ledger(account.clone())
			.or_else(|e| Err(e.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))))?;
```

**Review by Ank4n — APPROVED** [review]
Really awesome work with the refactor. 😇

**Review by kianenigma — APPROVED** [review]

**Review by sam0x17 — APPROVED** [review]

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/recover-corrupted-staking-ledgers-in-polkadot-and-kusama/9796/1

**Pull request merged**